src/server/server.service.js

```diff
@@ -15,6 +19,171 @@ module.exports = class Server {
+        } catch (err) {
+            console.error("Failed to connect to the DB: " + err);
+        }
+
+        var offset = this.headlines_offset[lang];
+        var limit = this.headlines_limit;
+
+        // fetch possible headlines
+        var q = 'SELECT headline,link ' +
+            `FROM ${lang}_headlines ` +
+            'ORDER BY id ' +
+            `LIMIT ${limit} ` +
+            `OFFSET ${offset} `;
+
+        await db.query(q)
+            .then(res => {
+                for (let row of res.rows) {
+                    debug(JSON.stringify(row));
+                }
+
+                if ((offset + limit) > this.headlines_count[lang]) {
+                    // If reached the end, go back to the beginning
+                    this.headlines_offset[lang] = 0;
+                } else {
+                    // Otherwise move the offset further
+                    this.headlines_offset[lang] = offset + limit;
+                }
+                this.headlines[lang] = this.headlines[lang].concat(this.shuffle(res.rows));
+
+                console.log(`Headlines for language ${lang} fetched`);
+            })
+            .catch(err => {
+                console.error(err);
+                throw err;
+            })
+            .finally(() => {
+                db.end();
+            });
+    }
+
+    async initHeadlines() {
+        var dbConfig;
+        if (process.env.NODE_ENV === 'production') {
+            dbConfig = {
+                connectionString: process.env.DATABASE_URL,
+                ssl: {
+                    sslmode: 'require',
+                    rejectUnauthorized: false,
+                }
+            };
+        } else {
+            dbConfig = {
+                connectionString: process.env.DATABASE_URL,
+            };
+        }
+
+        const db = new Client(dbConfig);
+
+        try {
+            await db.connect();
+        } catch (err) {
+            console.error("Failed to connect to the DB: " + err);
+        }
+
+        var limit = this.headlines_limit;
+        var promises = [];
+        for (let lang of consts.SUPPORTED_LANGUAGES) {
+            //this.headlines_mutex[lang] = new Mutex();
+            let q = `SELECT COUNT(id) FROM ${lang}_headlines`;
+            promises.push(db.query(q)
+                .then(async (res) => {
+                    let count = parseInt(res.rows[0]['count']);
+                    this.headlines_count[lang] = count;
+
+                    // If the number of rows is larger than the limit, choose
+                    // starting offset randomly. Otherwise get all rows.
+                    if (count > limit) {
+                        this.headlines_offset[lang] = Math.floor(Math.random() *
+                            (count - limit));
+                    } else {
+                        this.headlines_offset[lang] = 0;
+                    }
+
+                    this.headlines[lang] = [];
+
+                    await this.fetchHeadlines(lang);
```

Comment:
`fetchHeadLines()` already connects to the DB, does it mean that `initHeadlines()` connects twice to the DB?

Reply:
The impact of this is probably low (once every time the server starts), but just wanted to check awareness of this behaviour.

Reply:
Ack, I knew about the double connection at startup.